### PR TITLE
Memoize class binding when compiling private methods and static elements

### LINKED ISSUE #15696: [Bug]: Incorrect transpilation of class name references in static contexts
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
class Foo {
  static { console.log(this, Foo) }
  static x = () => { console.log(this, Foo) }
}
const oldFoo = Foo;
Foo = null;
oldFoo.x();
```

### Configuration file name

_No response_

### Configuration

(I put the input code into the [Babel REPL](https://babel.dev/repl))

### Current and expected behavior

Current behavior: When the output code Babel generates is run, it prints something like this:

```
Foo Foo
null null
```

When the input code is run natively, it prints something like this:

```
Foo Foo
Foo Foo
```

I expected the behavior of the output code Babel generates to match behavior of the input code.

### Environment

(I put the input code into the [Babel REPL](https://babel.dev/repl))

### Possible solution

Class declarations in JavaScript generate an outer binding name, which turns out to be mutable (and can be reassigned). Separately, JavaScript also generates an immutable inner binding name that is only in scope within the class body. Referencing the class name within the class body gives the inner immutable binding name, and referencing the class name outside the class body gives the mutable outer binding name.

Currently Babel generates this output code:

```js
var Foo = /*#__PURE__*/_createClass(function Foo() {
  _classCallCheck(this, Foo);
});
console.log(Foo, Foo);
_defineProperty(Foo, "x", function () {
  console.log(Foo, Foo);
});
var oldFoo = Foo;
Foo = null;
oldFoo.x();
```

This problem could be fixed by generating the following output code instead:

```js
var Foo = /*#__PURE__*/_createClass(function Foo() {
  _classCallCheck(this, Foo);
});
var _Foo = Foo;
console.log(_Foo, _Foo);
_defineProperty(_Foo, "x", function () {
  console.log(_Foo, _Foo);
});
var oldFoo = Foo;
Foo = null;
oldFoo.x();
```

The modified code explicitly materializes the immutable inner binding name for the code that had to be moved outside of the class body due to syntax lowering.

### Additional context

I discovered this bug in Babel because I'm trying to figure out how to handle this in esbuild, and as part of that I'm investigating how Babel transpiles code that does this.

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #15696 
| Patch: Bug Fix?          | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

In this PR we memoise the class binding when it is referenced in private methods or static elements, either via `this` or the class id. Previously we replaced `this` with class id, but as #15696 point out,

> Class declarations in JavaScript generate an outer binding name, which turns out to be mutable (and can be reassigned). Separately, JavaScript also generates an immutable inner binding name that is only in scope within the class body.

it will fail when the outer binding is reassigned.

I suggest reviewing this PR by commits as there are many fixture changes.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15701"></a>

### CODE REVIEW

File: packages/babel-helper-create-class-features-plugin/src/fields.ts [hunk 1]
```diff
@@ -1103,6 +1110,10 @@ export function buildFieldsInitNodes(
         return injectSuperRef;
       };
 
+  const classRefForInnerBinding =
```
Comment (JLHwung) [inline packages/babel-helper-create-class-features-plugin/src/fields.ts:1113]: When `ref` is not defined and `classRefFlags.ForInnerBinding` is not enabled, e.g. compiling a plain class `class C {}` without private methods / static elements, we will generate a uid identifier but we do not insert it to the final AST. In this case the memoizer variable `_class0`, `_class1`, etc. is not continuous.

**Comment by nicolo-ribaudo** [comment]
Is only generating the extra ref when the class is reassigned doable?

**Comment by nicolo-ribaudo** [comment]
I  pushed some commits to do https://github.com/babel/babel/pull/15701#issuecomment-1597547398, wdyt about this approach?

**Comment by JLHwung** [comment]
@nicolo-ribaudo Thanks for working on this. I didn't wrap it under the reassignment check because 
- It generally does not work with the case where the reassignment is introduced by a plugin. The replacement methods like `NodePath#insertAfter` or `NodePath#replaceWith` will not requeue the binding node. And it's asking a bit too much for plugin authors to do so.
- The extra assignment does not necessarily bloat the output size. When the class id is long and class bindings are frequently used, the reassignment is doing a favor for the compressor as all the reference identifiers are now able to be mangled. Because of the compressor I didn't go ahead and fix the uuid continuity. 

Btw TypeScript memoizes the class binding without checking if it is reassigned, too: https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG8BQ0PQgFzNglsNAMQjQC8ciA3KgL5A

**Review by nicolo-ribaudo — APPROVED** [review]
👍 I reverted my changes, the code is good

**Review by JLHwung — COMMENTED** [review]
✅ Just rebased this PR and it looks good to me.

**Pull request merged**